web/modules/custom/joinup_notification/src/JoinupMessageDelivery.php

```diff
@@ -0,0 +1,145 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup_notification;
+
+use Drupal\message\Entity\Message;
+use Drupal\message\MessageInterface;
+use Drupal\message_notify\MessageNotifier;
+use Drupal\user\UserInterface;
+
+/**
+ * Provides a service class for creating and delivering messages.
+ */
+class JoinupMessageDelivery implements JoinupMessageDeliveryInterface {
+
+  /**
+   * The message to be delivered.
+   *
+   * @var \Drupal\message\MessageInterface
+   */
+  protected $message;
+
+  /**
+   * A list of user accounts acting as recipients for the message.
+   *
+   * @var \Drupal\user\UserInterface[]
+   */
+  protected $accounts = [];
+
+  /**
+   * E-mail recipients.
+   *
+   * @var string[]
+   */
+  protected $mails = [];
+
+  /**
+   * The message notifier service.
+   *
+   * @var \Drupal\message_notify\MessageNotifier
+   */
+  protected $messageNotifier;
+
+  /**
+   * Additional arguments.
+   *
+   * @var array
+   */
+  protected $arguments = [];
+
+  /**
+   * Constructs a new Joinup deliver service object.
+   *
+   * @param \Drupal\message_notify\MessageNotifier $message_notifier
+   *   The message notifier service.
+   */
+  public function __construct(MessageNotifier $message_notifier) {
+    $this->messageNotifier = $message_notifier;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setMessage(MessageInterface $message): JoinupMessageDeliveryInterface {
+    $this->message = $message;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function createMessage(string $message_template, array $values = []): JoinupMessageDeliveryInterface {
+    // If the template was passed in $values, $message_template take precedence.
+    $values = ['template' => $message_template] + $values;
+    $this->message = Message::create($values);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setArguments(array $arguments): JoinupMessageDeliveryInterface {
+    $this->arguments = $arguments;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setRecipients(array $accounts): JoinupMessageDeliveryInterface {
+    $this->accounts = $accounts;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setRecipientsAsEmails(array $mails): JoinupMessageDeliveryInterface {
+    $this->mails = $mails;
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function sendMail(): JoinupMessageDeliveryInterface {
```

Comment:
Since this is part of a service this shouldn't rely on data which is stored in properties on the object. The properties might be set during a previous usage of the service and may be outdated.

I made a followup to fix this: ISAICP-4169.